DataStore documentation revamp (Getting Started) Part 1/2 (#1827)
aws-amplify/docs: docs/lib/datastore/getting-started.md

---
title: Getting started
description: The Amplify DataStore category enables developers to write applications whose data lives in the mobile device for instant/offline-first access while synchronization with a cloud backend is handled automatically.
---

> The Amplify DataStore category enables developers to write applications whose data lives in the mobile device for instant/offline-first access while synchronization with a cloud backend powered by [AppSync](https://aws.amazon.com/appsync/) is handled automatically, which makes working with distributed, cross-user data just as simple as working with local-only data.

Author: We talked offline but will register here so other folks can chime in.

I personally find the existing description you mentioned confusing and *buzz-wordy*, but I acknowledge I may have failed to simplify it.

I don't think *"programming model for leveraging shared and distributed data"* is how we should introduce what DataStore is, maybe that suits more senior developers expectations, but considering developers from all levels DataStore is a database with offline and cloud-sync capabilities, how to present that in simple terms?

> The Amplify DataStore category enables developers to write applications whose data lives in the mobile device for instant/offline-first access while synchronization with a cloud backend powered by [AppSync](https://aws.amazon.com/appsync/) is handled automatically, which makes working with distributed, cross-user data just as simple as working with local-only data.

Author: By popular demand and lack of a better option I reverted to the original introduction paragraph.

```console
? Please select from one of the below mentioned services:
    `GraphQL`
? Choose the default authorization type for the API
    `API key`
? Enter a description for the API key:
    `defaultapikey`
? After how many days from now the API key should expire (1-365):
    `365`
? Do you want to configure advanced settings for the GraphQL API
    `Yes, I want to make some additional changes.`
? Configure additional auth types?
    `No`
? Configure conflict detection?
    `Yes`
? Select the default resolution strategy
    `Auto Merge`
? Do you want to override default per model settings?
    `No`
```

Generate the models by running in the terminal.

```console
amplify codegen models
```

The files should be generated at `amplify/generated/models/`, add them to your Xcode project (TODO elaborate on how to do it?).

<amplify-callout>

> **Troubleshooting:** without the **conflict detection** configuration cloud sync will fail. If that's the case use `amplify update api` and choose **Enable DataStore for entire API** (this option will enable the conflict detection as described above).

Reviewer: amplify-callout warning?

</amplify-callout>

> ## Initialize Amplify Storage

Reviewer: Storage?

Author: a small leftover :)
fixed already

---
title: Getting started
description: The Amplify DataStore category enables developers to write applications whose data lives in the mobile device for instant/offline-first access while synchronization with a cloud backend is handled automatically.
---

> The Amplify DataStore category enables developers to write applications whose data lives in the mobile device for instant/offline-first access while synchronization with a cloud backend powered by [AWS AppSync](https://aws.amazon.com/appsync/) is handled automatically, which makes working with distributed, cross-user data just as simple as working with local-only data.

Reviewer 2: This is a bit of a run-on sentence.  I feel like we should lead in with the problem that developers face when... the wording needs to be updated, but something like:
Handling data in a modern mobile application is difficult due to data constantly changing coupled with poor network conditions.  Meanwhile, users increasingly demand "snappy" applications which has lead to techniques like optimistic UI which assume network calls will eventually succeed.  The Amplify DataStore category enables developers to adopt a programming model where data lives on device so that it is instantly available even when customers have a poor internet connection.  Furthermore, ... etc.. etc..

I'm also fine to just leave it as you have it right now, and update it at a later time.

Author: I'm open to suggestions but I have a strong opinion that the current content doesn't tell me what DataStore is. I can only understand the current paragraph if I open that page knowing that I'm looking at a offline/cloud-sync database solution.

When I look at the other categories I can easily understand what they do right away.

Is there a way to describe DataStore in a way someone who only used CoreData or text files to persist data would understand?

Author: By popular demand and lack of a better option I reverted to the original introduction paragraph.

<amplify-callout>

**Note:** cloud synchronization is optional and disabled by default. This allows you to start with Amplify DataStore right away, even without an AWS account.

</amplify-callout>

## Goal
To setup and configure your application with Amplify DataStore and use it to persist data locally on a device.

## Prerequisites

<inline-fragment platform="ios" src="~/lib/datastore/fragments/ios/getting-started/10_preReq.md"></inline-fragment>
<inline-fragment platform="android" src="~/lib/datastore/fragments/android/getting-started/10_preReq.md"></inline-fragment>

## Data schema

> The first step to create an app backed by a persistent datastore is to **define a schema**. DataStore uses GraphQL schema files as the the definition of the application data model. The schema contains data types and relationships that represent the app functionality.

Reviewer 2: replace "as the the" with "as the"
replace "app" with "app's"

### Sample schema

Your first step is to define a schema for a small blog application. It has a single model, a `Post`. New types and constructs will be added to this base schema as more concepts are presented.

> Open the `schema.graphql` file located by default at `amplify/backend/{api_name}/`  and **define the model** `Post` as follows.

Reviewer 2: At the time customers are seeing this, schema.graphql does not exist..  Should we have them create a directory and create this file?

Author: hhmm good call! I simplified the steps a bit too much, you're right the build needs to happen once so we can have the `schema.graphql` created. I'll shift things around and see how I can keep it simple and consistent with the required steps

Author: fixed, much better now!

In order to start choose an option to **generate models** from the `schema.graphql`.

### Option 1: Use IDE integration

<inline-fragment platform="ios" src="~/lib/datastore/fragments/ios/getting-started/30_amplifyTools.md"></inline-fragment>
<inline-fragment platform="android" src="~/lib/datastore/fragments/android/getting-started/30_amplifyTools.md"></inline-fragment>

### Option 2: Use Amplify CLI

If more control over the configuration and code generation process is needed the Amplify CLI can be used directly. This option is also useful for **existing codebases** with Amplify already configured.

The base structure for a DataStore app is created by adding a new GraphQL API to your app.

```console
amplify add api
```

During the API configuration process select **GraphQL** as the API type and reply to the questions as follows. Make sure you respond **yes** to **advanced settings** and turn on **conflict detection**. This setting is critical when syncing data to the cloud since the conflict resolution strategy is what allows local data to be reconciled with data from the cloud backend.

```console
? Please select from one of the below mentioned services:
    `GraphQL`

> ? Choose the default authorization type for the API

Reviewer 2: Seemsl ike we forgot a line, i think this should be:
```
? Please select from one of the below mentioned services:
    `GraphQL`
? Provide API name:
    `mydatastoreapp`
? Choose the default authorization type for the API
    `API key`

... then things.. go off the rails here:

? Select the default resolution strategy
   Auto Merge
? Do you have an annotated GraphQL schema?
   No
? Do you want a guided schema creation?
   No
? Provide a custom type name (MyType)
   ???
```

Author: fixed

### Writing to the database

To write to the database, create an instance of the `Post` model and save it.

<inline-fragment platform="ios" src="~/lib/datastore/fragments/ios/getting-started/50_saveSnippet.md"></inline-fragment>
<inline-fragment platform="android" src="~/lib/datastore/fragments/android/getting-started/50_saveSnippet.md"></inline-fragment>

### Reading from the database

To read from the database, the simplest approach is to query for all records of a given model type.

<inline-fragment platform="ios" src="~/lib/datastore/fragments/ios/getting-started/60_querySnippet.md"></inline-fragment>
<inline-fragment platform="android" src="~/lib/datastore/fragments/android/getting-started/60_querySnippet.md"></inline-fragment>

## Next steps

Congratulations! You’ve created and retrieved data from the local database. Check out the following links to see other Amplify DataStore use cases and advanced concepts:

- Queries with conditions and pagination
- Delete records
- Cloud synchronization
- Model associations

> - Clear the local database

Reviewer 2: we should probably add links for these?

Author: will do

<amplify-callout>

> **Note:** cloud synchronization is optional and disabled by default. This allows you to start with DataStore right away, even without an AWS account.

Reviewer 2: * Not sure who updated it but, I love the intro where it says "Amplify DataStore provides a programming model for leveraging shared..."

* This allows you to start persisting data locally to your device with DataStore, even without an AWS account.

</amplify-callout>

## Goal
To setup and configure your application with Amplify DataStore and use it to persist data locally on a device.

## Prerequisites

<inline-fragment platform="js" src="~/lib/datastore/fragments/js/getting-started/10_preReq.md"></inline-fragment>
<inline-fragment platform="ios" src="~/lib/datastore/fragments/ios/getting-started/10_preReq.md"></inline-fragment>
<inline-fragment platform="android" src="~/lib/datastore/fragments/android/getting-started/10_preReq.md"></inline-fragment>

## Install Amplify Libraries

<inline-fragment platform="ios" src="~/lib/datastore/fragments/ios/getting-started/20_installLib.md"></inline-fragment>
<inline-fragment platform="android" src="~/lib/datastore/fragments/android/getting-started/20_installLib.md"></inline-fragment>

There are two options to integrate the Amplify build process with the project.

> 

Reviewer 2: I think there is a potential problem here.  In our pre-req, we have them go through a project setup, which has them manually adding amplifyconfiguration.json and awsconfiguration.json into the project.  This seems to conflict with the instructions here...

option 1:
* put this option1/option2 thing into the project setup
* then assume they have done one or the other here, and update the verbage

Option 2:
* do not reference amplify tools in this guide.
* leave project setup as-is

Option 3:
* hmmm something else?

Reviewer 2: Fixing this some other time, resolving.

### Option 1: Platform integration

<inline-fragment platform="js" src="~/lib/datastore/fragments/js/getting-started/30_platformIntegration.md"></inline-fragment>
<inline-fragment platform="ios" src="~/lib/datastore/fragments/ios/getting-started/30_platformIntegration.md"></inline-fragment>
<inline-fragment platform="android" src="~/lib/datastore/fragments/android/getting-started/30_platformIntegration.md"></inline-fragment>

> ### Option 2: Use Amplify CLI

Reviewer 3: Remove this option for Android - we should just present one option for now.

> If more control over the configuration and code generation process is needed the Amplify CLI can be used directly. This option is particularly useful for **existing codebases** where Amplify is already configured.

Reviewer 2: Saying "if more control over configuration" ... seems vague to me, because it doesn't call out an example.  How about something like:

Instead of integrating Amplify Tools into your IDE, you can alternatively use the Amplify CLI on its own to accomplish the same thing that Amplify Tools is doing for you.  This option is particularly useful for .....

I think this flows better because it calls out what the previous option is and and then uses the word "alternatively" to put both options on equal footing.

Author: Nice, I like it!

The base structure for a DataStore app is created by adding a new GraphQL API to your app.

```console
amplify add api
```

During the API configuration process select **GraphQL** as the API type and reply to the questions as follows. Make sure you respond **Yes, I want to make some additional changes** when prompted for **advanced settings** and turn on **conflict detection**. This setting is critical when syncing data to the cloud since the conflict resolution strategy is what allows local data to be reconciled with data from the cloud backend.

```console
? Please select from one of the below mentioned services:
    `GraphQL`

> ? Provide API name:
>     `BlogAppApi`
> ? Choose the default authorization type for the API
>     `API key`
> ? Enter a description for the API key:
>     `BlogAPIKey`

Reviewer 4: ```suggestion
? Provide API name:
    `PostAppApi`
? Choose the default authorization type for the API
    `API key`
? Enter a description for the API key:
    `PostAPIKey`
```

? After how many days from now the API key should expire (1-365):
    `365`
? Do you want to configure advanced settings for the GraphQL API
    `Yes, I want to make some additional changes.`
? Configure additional auth types?
    `No`
? Configure conflict detection?
    `Yes`
? Select the default resolution strategy
    `Auto Merge`
? Do you have an annotated GraphQL schema?
    `No`
? Do you want a guided schema creation?
    `No`
? Provide a custom type name
    `Post`
```

<amplify-callout warning>

> **Troubleshooting:** without the **conflict detection** configuration cloud sync will fail. In that case use `amplify update api` and choose **Enable DataStore for entire API** (this option will enable the conflict detection as described above).

Reviewer 2: AWESOME

</amplify-callout>

## Idiomatic persistence

DataStore relies on platform standard data structures to represent the data schema in an idiomatic way. The persistence language is composed by data types that satisfies the `Model` interface and operations defined by common verbs such as `save`, `query` and `delete`.

### Data schema

The first step to create an app backed by a persistent datastore is to **define a schema**. DataStore uses GraphQL schema files as the definition of the application data model. The schema contains data types and relationships that represent the app's functionality.

### Sample schema

For the next steps, let's start with a schema for a small blog application. It has a single model, a `Post`. New types and constructs will be added to this base schema as more concepts are presented.

> Open the `schema.graphql` file located by default at `amplify/backend/{api_name}/` and **define the model** `Post` as follows.

Reviewer: ```suggestion
Open the `schema.graphql` file located by default at `amplify/backend/api/amplifyDatasource` and **define the model** `Post` as follows.
```

Reviewer 2: I think the location where we are saving schema.graphql has changed in the latest version of the CLI.  Daniel's instructions seem correct to me.

Author: it really depends on what method a developer use for creating the API. The `amplifyDatasource` is true only for `amplify-app`, devs using `amplify add api` can pick any name, that's why I went with `{api_name}`.

Thoughts @wooj2 @jpignata ?

Reviewer 2: {api_name} sounds good to me

```graphql

> type Post @model {

Reviewer 3: This schema is inconsistent with the schema used for the subsequent sections (see Manipulating data, there is a status field which refers to a PostStatus enum.

Reviewer 4: i just noticed this as i was going through it

  id: ID!
  title: String!
  rating: Int
  content: String
}
```

Now you will to convert the platform-agnostic `schema.graphql` into platform-specific data structures that will enforce the idiomatic persistence concepts. DataStore relies on code generation to guarantee schemas are correctly converted to platform code.

> Like the initial setup, models can be generated either using the IDE integration or Amplify CLI directly.

Reviewer 2: This sentence is unclear.  Seems like we can just remove it?

### Code generation: Platform integration

<inline-fragment platform="js" src="~/lib/datastore/fragments/js/getting-started/40_codegen.md"></inline-fragment>
<inline-fragment platform="ios" src="~/lib/datastore/fragments/ios/getting-started/40_codegen.md"></inline-fragment>
<inline-fragment platform="android" src="~/lib/datastore/fragments/android/getting-started/40_codegen.md"></inline-fragment>

> ### Code generation: Amplify CLI

Reviewer 3: This section shouldn't be present for Android - also it specifically mentions Xcode below.

> Models can also be generated using the Amplify CLI directly.

Reviewer 2: Suggestion, i know this makes it harder since we have to split up each platform, but i think it makes it more clear:

If you have chosen to not use AmplifyTools or want to generate models without issuing a build in XCode, you can use the following steps:

Reviewer 2: I guess we are leaving this to do for another time...

> 1. In your terminal **execute the codegen command**:

Reviewer 2: Suggestion:
1. In your terminal, change directories to your project's folder and **execute the codegen command**:

### Writing to the database

To write to the database, create an instance of the `Post` model and save it.

<inline-fragment platform="js" src="~/lib/datastore/fragments/js/getting-started/60_saveSnippet.md"></inline-fragment>
<inline-fragment platform="ios" src="~/lib/datastore/fragments/ios/getting-started/60_saveSnippet.md"></inline-fragment>
<inline-fragment platform="android" src="~/lib/datastore/fragments/android/getting-started/60_saveSnippet.md"></inline-fragment>

### Reading from the database

To read from the database, the simplest approach is to query for all records of a given model type.

<inline-fragment platform="js" src="~/lib/datastore/fragments/js/getting-started/70_querySnippet.md"></inline-fragment>
<inline-fragment platform="ios" src="~/lib/datastore/fragments/ios/getting-started/70_querySnippet.md"></inline-fragment>
<inline-fragment platform="android" src="~/lib/datastore/fragments/android/getting-started/70_querySnippet.md"></inline-fragment>

## Next steps

Congratulations! You’ve created and retrieved data from the local database. Check out the following links to see other Amplify DataStore use cases and advanced concepts:

> - [Write data](~/lib/datastore/data-access.md#insert-and-update)

Reviewer 4: `[Manipulating data]` to match the section heading?

Reviewer 4: ```suggestion
- [Write data](~/lib/datastore/data-access.md#save-date)
```

<amplify-callout>

**Note:** this allows you to start persisting data locally to your device with DataStore, even without an AWS account.

</amplify-callout>

## Goal
To setup and configure your application with Amplify DataStore and use it to persist data locally on a device.

## Prerequisites

<inline-fragment platform="js" src="~/lib/datastore/fragments/js/getting-started/10_preReq.md"></inline-fragment>
<inline-fragment platform="ios" src="~/lib/datastore/fragments/ios/getting-started/10_preReq.md"></inline-fragment>
<inline-fragment platform="android" src="~/lib/datastore/fragments/android/getting-started/10_preReq.md"></inline-fragment>

> 

Reviewer 5: Could be that React Native case is not be completely clear here?

Author: Is there a special setup process for RN?

<inline-fragment platform="ios" src="~/lib/datastore/fragments/ios/getting-started/20_installLib.md"></inline-fragment>
<inline-fragment platform="android" src="~/lib/datastore/fragments/android/getting-started/20_installLib.md"></inline-fragment>

There are two options to integrate the Amplify build process with the project.

### Option 1: Platform integration

<inline-fragment platform="js" src="~/lib/datastore/fragments/js/getting-started/30_platformIntegration.md"></inline-fragment>
<inline-fragment platform="ios" src="~/lib/datastore/fragments/ios/getting-started/30_platformIntegration.md"></inline-fragment>
<inline-fragment platform="android" src="~/lib/datastore/fragments/android/getting-started/30_platformIntegration.md"></inline-fragment>

### Option 2: Use Amplify CLI

Instead of using the platform integration, you can alternatively use the Amplify CLI on its own to accomplish the same thing that Amplify Tools is doing for you. This option is particularly useful for **existing projects** where Amplify is already configured.

The base structure for a DataStore app is created by adding a new GraphQL API to your app.

```console
amplify add api
```

> During the API configuration process select **GraphQL** as the API type and reply to the questions as follows. Make sure you respond **Yes, I want to make some additional changes** when prompted for **advanced settings** and turn on **conflict detection**. This setting is critical when syncing data to the cloud since the conflict resolution strategy is what allows local data to be reconciled with data from the cloud backend.

Reviewer 5: I think is more **required** than critical. Because that turn on DataStore on the backend and generates the model for the clients.

```console
? Please select from one of the below mentioned services:
    `GraphQL`
? Provide API name:
    `BlogAppApi`
? Choose the default authorization type for the API
    `API key`
? Enter a description for the API key:
    `BlogAPIKey`
? After how many days from now the API key should expire (1-365):
    `365`
? Do you want to configure advanced settings for the GraphQL API
    `Yes, I want to make some additional changes.`
? Configure additional auth types?
    `No`
? Configure conflict detection?
    `Yes`
? Select the default resolution strategy
    `Auto Merge`
? Do you have an annotated GraphQL schema?
    `No`
? Do you want a guided schema creation?
    `No`
? Provide a custom type name
    `Post`
```

<amplify-callout warning>

> **Troubleshooting:** without the **conflict detection** configuration cloud sync will fail. In that case use `amplify update api` and choose **Enable DataStore for entire API** (this option will enable the conflict detection as described above).

Reviewer 5: I would make the previous reference to this more obvious. Is strange that **Enable DataStore for entire API** is only on `api update` and not on `api add` command

Author: Yeah, I agree that that question being only present in the update flow is weird.

Can you elaborate on *"I would make the previous reference to this more obvious"*?

</amplify-callout>

## Idiomatic persistence

DataStore relies on platform standard data structures to represent the data schema in an idiomatic way. The persistence language is composed by data types that satisfies the `Model` interface and operations defined by common verbs such as `save`, `query` and `delete`.

### Data schema

The first step to create an app backed by a persistent datastore is to **define a schema**. DataStore uses GraphQL schema files as the definition of the application data model. The schema contains data types and relationships that represent the app's functionality.

### Sample schema

For the next steps, let's start with a schema for a small blog application. It has a single model, a `Post`. New types and constructs will be added to this base schema as more concepts are presented.

Open the `schema.graphql` file located by default at `amplify/backend/{api_name}/` and **define the model** `Post` as follows.

```graphql
type Post @model {
  id: ID!
  title: String!
  status: PostStatus!
  rating: Int
  content: String
}

> enum PostStatus {

Reviewer 6: Is the enum necessary for a getting started or should be make this as basic as possible?

Author: @dabit3 that a very good callout, but tbh as much as I want to make it as simple as possible, right now we lack a more organized page on how to define schemas and all the capabilities for newcomers of DataStore. We do have a "transformer" documentation outside of the DataStore scope but that not only takes the developer to another page but also into a much deeper GraphQL context than what DataStore is trying to expose.

But I would say we should take that as a goal on the next improvement we need to add to DataStore docs: a schema mapping page, tools like Hibernate and Rails ActiveRecord invest heavily into document best practices and different scenarios when defining a data schema (your recipes repo could serve as a base I think!).

### Code generation: Amplify CLI

Models can also be generated using the Amplify CLI directly.

1. In your terminal, change directories to your project's folder and **execute the codegen command**:
    ```console
    amplify codegen models
    ```
2. **Locate the generated files** at `amplify/generated/models/`.
3. **Add the files** to the Xcode project.

## Initialize Amplify DataStore

<inline-fragment platform="js" src="~/lib/datastore/fragments/js/getting-started/50_initDataStore.md"></inline-fragment>
<inline-fragment platform="ios" src="~/lib/datastore/fragments/ios/getting-started/50_initDataStore.md"></inline-fragment>
<inline-fragment platform="android" src="~/lib/datastore/fragments/android/getting-started/50_initDataStore.md"></inline-fragment>

## Persistence operations

Now the application is ready to execute persistence operations. The data will be persisted to a local database, enabling offline-first use cases by default.

Even though a GraphQL API is already added to your project, the cloud synchronization will only be enabled when the API plugin is initialized and the backend provisioned. See the [Next steps](#next-steps) for more info.

> ### Writing to the database

Reviewer 5: Is local store more accurate than database?
